File: src/infrastructure/tools/file/mkdir.ts
/**
 * mkdir — 创建目录（支持单个 path 与批量 paths，幂等：已存在视为成功）
 */
import * as tauriFs from '@tauri-apps/plugin-fs'
import { t } from '@/ui/i18n'
import { toolRegistry } from '@/domain/tools'
import type { ToolContext, ToolExecutor, ToolResult } from '@/domain/tools/types'
import { securityService } from '@/services/security-service'
import { isTauriFsAvailable } from './common'

toolRegistry.register(
    'mkdir',
    (async (args: Record<string, any>, ctx: ToolContext): Promise<ToolResult> => {
    if (!isTauriFsAvailable())
      throw '[mkdir] Error: not running in a Tauri environment'

    const recursive = args.recursive !== false

    // 兼容单个 path 与多个 paths；过滤空字符串
    const rawPaths: string[] = Array.isArray(args.paths)
      ? (args.paths as any[]).filter(
          (p): p is string => typeof p === 'string' && p.trim() !== '',
        )
      : typeof args.path === 'string' && args.path.trim() !== ''
        ? [args.path]
        : []

    if (rawPaths.length === 0) {
      throw 'Error: provide either the "path" or the "paths" parameter'
    }

    const created: string[] = []
    const existed: string[] = []
    const errors: string[] = []

    for (const raw of rawPaths) {
      const fullPath = await securityService.resolveSafePath(
        raw,
        'w',
        ctx.sessionId,
      )
      try {
        const alreadyExists = await tauriFs.exists(fullPath)
        if (alreadyExists) {
          existed.push(fullPath)
          continue
        }
        await tauriFs.mkdir(fullPath, { recursive })
        created.push(fullPath)
      } catch (e: any) {
        errors.push(`${fullPath} — ${e.message || String(e)}`)
      }
    }

    // ⚠️ 模型侧固定英文（P4b/D2-A）；UI 侧走 uiData。文案与 Rust
    // `native_tools/file/mkdir.rs` 逐字一致（铁律 1）。
    const parts: string[] = []
    if (created.length > 0) {
      parts.push(
        created.length === 1
          ? `📁 Directory created: ${created[0]}`
          : `📁 Created ${created.length} directories:\n${created
              .map((p) => `  - ${p}`)
              .join('\n')}`,
      )
    }
    if (existed.length > 0) {
      parts.push(
        existed.length === 1
          ? `ℹ️ Directory already exists: ${existed[0]}`
          : `ℹ️ ${existed.length} directories already exist:\n${existed
              .map((p) => `  - ${p}`)
              .join('\n')}`,
      )
    }
    if (errors.length > 0) {
      parts.push(
        `⚠️ Failed to create ${errors.length} directories:\n${errors
          .map((e) => `  - ${e}`)
          .join('\n')}`,
      )
    }

    if (created.length === 0 && existed.length === 0 && errors.length > 0) {
      throw errors.join('\n')
    }

    return {
      content: parts.join('\n'),
      uiData: {
        created,
        existed,
        errors,
      },
    }
  }) as ToolExecutor,
    t('创建目录'),
)

File: src-tauri/src/agent/native_tools/file/mkdir.rs
//! `mkdir` 工具（原生）— 创建目录，支持 `path` 或 `paths`、recursive 开关；已存在的计入 existed。

use crate::agent::native_tools::common::{arg_bool, arg_str, resolve_safe_path};
use crate::agent::native_tools::{NativeToolCtx, NativeToolOutcome};
use serde_json::{json, Value};

pub(crate) async fn mkdir_tool(
    ctx: &NativeToolCtx<'_>,
    args: &Value,
) -> Result<NativeToolOutcome, String> {
    let recursive = arg_bool(args, "recursive").unwrap_or(true);

    // 兼容单个 path 与多个 paths
    let mut raw_paths: Vec<String> = Vec::new();
    if let Some(arr) = args.get("paths").and_then(|v| v.as_array()) {
        for v in arr {
            if let Some(s) = v.as_str() {
                let s = s.trim();
                if !s.is_empty() {
                    raw_paths.push(s.to_string());
                }
            }
        }
    }
    if raw_paths.is_empty() {
        if let Some(p) = arg_str(args, "path") {
            let p = p.trim();
            if !p.is_empty() {
                raw_paths.push(p.to_string());
            }
        }
    }

    if raw_paths.is_empty() {
        return Err("错误：请提供 \"path\" 或 \"paths\" 参数".to_string());
    }

    let mut created: Vec<String> = Vec::new();
    let mut existed: Vec<String> = Vec::new();
    let mut errors: Vec<String> = Vec::new();

    for raw in &raw_paths {
        let full_path = resolve_safe_path(raw, "w", ctx.security)?;
        let path = std::path::Path::new(&full_path);

        if path.exists() {
            existed.push(full_path);
            continue;
        }

        let result = if recursive {
            std::fs::create_dir_all(path)
        } else {
            std::fs::create_dir(path)
        };

        match result {
            Ok(()) => created.push(full_path),
            Err(e) => errors.push(format!("{} — {}", full_path, e)),
        }
    }

    let mut parts: Vec<String> = Vec::new();
    if !created.is_empty() {
        if created.len() == 1 {
            parts.push(format!("📁 已创建目录: {}", created[0]));
        } else {
            parts.push(format!(
                "📁 已创建 {} 个目录:\n{}",
                created.len(),
                created.iter().map(|p| format!("  - {}", p)).collect::<Vec<_>>().join("\n")
            ));
        }
    }
    if !existed.is_empty() {
        if existed.len() == 1 {
            parts.push(format!("ℹ️ 目录已存在: {}", existed[0]));
        } else {
            parts.push(format!(
                "ℹ️ 已存在 {} 个目录:\n{}",
                existed.len(),
                existed.iter().map(|p| format!("  - {}", p)).collect::<Vec<_>>().join("\n")
            ));
        }
    }
    if !errors.is_empty() {
        parts.push(format!(
            "⚠️ 有 {} 个目录创建失败:\n{}",
            errors.len(),
            errors.iter().map(|e| format!("  - {}", e)).collect::<Vec<_>>().join("\n")
        ));
    }

    if created.is_empty() && existed.is_empty() && !errors.is_empty() {
        return Err(errors.join("\n"));
    }

    Ok(NativeToolOutcome::Value {
        content: parts.join("\n"),
        ui_data: Some(json!({
            "created": created,
            "existed": existed,
            "errors": errors,
        })),
    })
}
